README
NAME
    DBIx::Class::Candy - Sugar for your favorite ORM, DBIx::Class

SYNOPSIS
     package MyApp::Schema::Result::Artist;

     use DBIx::Class::Candy -autotable => v1;

     primary_column id => {
       data_type => 'int',
       is_auto_increment => 1,
     };

     column name => {
       data_type => 'varchar',
       size => 25,
       is_nullable => 1,
     };

     has_many albums => 'A::Schema::Result::Album', 'artist_id';

     1;

DESCRIPTION
    "DBIx::Class::Candy" is a simple sugar layer for definition of
    DBIx::Class results. Note that it may later be expanded to add sugar for
    more "DBIx::Class" related things. By default "DBIx::Class::Candy":

    *   turns on strict and warnings

    *   sets your parent class

    *   exports a bunch of the package methods that you normally use to
        define your DBIx::Class results

    *   makes a few aliases to make some of the original method names a
        shorter or more clear

    *   defines very few new subroutines that transform the arguments passed
        to them

    It assumes a DBIx::Class::Core-like API, but you can tailor it to suit
    your needs.

IMPORT OPTIONS
    See "SETTING DEFAULT IMPORT OPTIONS" for information on setting these
    schema wide.

  -base
     use DBIx::Class::Candy -base => 'MyApp::Schema::Result';

    The first thing you can do to customize your usage of
    "DBIx::Class::Candy" is change the parent class. Do that by using the
    "-base" import option.

  -autotable
     use DBIx::Class::Candy -autotable => v1;

    Don't waste your precious keystrokes typing "table 'buildings'", let
    "DBIx::Class::Candy" do that for you! See "AUTOTABLE VERSIONS" for what
    the existing versions will generate for you.

  -components
     use DBIx::Class::Candy -components => ['FilterColumn'];

    "DBIx::Class::Candy" allows you to set which components you are using at
    import time so that the components can define their own sugar to export
    as well. See DBIx::Class::Candy::Exports for details on how that works.

  -perl5
     use DBIx::Class::Candy -perl5 => v10;

    I love the new features in Perl 5.10 and 5.12, so I felt that it would
    be nice to remove the boiler plate of doing "use feature ':5.10'" and
    add it to my sugar importer. Feel free not to use this.

IMPORTED SUBROUTINES
    Most of the imported subroutines are the same as what you get when you
    use the normal interface for result definition: they have the same names
    and take the same arguments. In general write the code the way you
    normally would, leaving out the "__PACKAGE__->" part. The following are
    methods that are exported with the same name and arguments:

     belongs_to
     has_many
     has_one
     inflate_column
     many_to_many
     might_have
     remove_column
     remove_columns
     resultset_attributes
     resultset_class
     sequence
     source_name
     table

    There are some exceptions though, which brings us to:

IMPORTED ALIASES
    These are merely renamed versions of the functions you know and love.
    The idea is to make your result classes a tiny bit prettier by aliasing
    some methods. If you know your "DBIx::Class" API you noticed that in the
    "SYNOPSIS" I used "column" instead of "add_columns" and "primary_key"
    instead of "set_primary_key". The old versions work, this is just nicer.
    A list of aliases are as follows:

     column            => 'add_columns',
     primary_key       => 'set_primary_key',
     unique_constraint => 'add_unique_constraint',
     relationship      => 'add_relationship',

SETTING DEFAULT IMPORT OPTIONS
    Eventually you will get tired of writing the following in every single
    one of your results:

     use DBIx::Class::Candy
       -base      => 'MyApp::Schema::Result',
       -perl5     => v12,
       -autotable => v1;

    You can set all of these for your whole schema if you define your own
    "Candy" subclass as follows:

     package MyApp::Schema::Candy;

     use base 'DBIx::Class::Candy';

     sub base { $_[1] || 'MyApp::Schema::Result' }
     sub perl_version { 12 }
     sub autotable { 1 }

    Note the "$_[1] ||" in "base". All of these methods are passed the
    values passed in from the arguments to the subclass, so you can either
    throw them away, honor them, die on usage, or whatever. To be clear, if
    you define your subclass, and someone uses it as follows:

     use MyApp::Schema::Candy -base => 'MyApp::Schema::Result', -perl5 => v18, -autotable => v1;

    Your "base" method will get "MyApp::Schema::Result", your "perl_version"
    will get 18, and your "autotable" will get 1.

SECONDARY API
  has_column
    There is currently a single "transformer" for "add_columns", so that
    people used to the Moose api will feel more at home. Note that this may
    go into a "Candy Component" at some point.

    Example usage:

     has_column foo => (
       data_type => 'varchar',
       size => 25,
       is_nullable => 1,
     );

  primary_column
    Another handy little feature that allows you to define a column and set
    it as the primary key in a single call:

     primary_column id => {
       data_type => 'int',
       is_auto_increment => 1,
     };

    If your table has multiple columns in its primary key, merely call this
    method for each column:

     primary_column person_id => { data_type => 'int' };
     primary_column friend_id => { data_type => 'int' };

  unique_column
    This allows you to define a column and set it as unique in a single
    call:

     unique_column name => {
       data_type => 'varchar',
       size => 30,
     };

AUTOTABLE VERSIONS
    Currently there is a single version, "v1", which looks at your class
    name, grabs everything after "::Schema::Result::" (or "::Result::"),
    removes the "::"'s, converts it to underscores instead of camel-case,
    and pluralizes it. Here are some examples if that's not clear:

     MyApp::Schema::Result::Cat -> cats
     MyApp::Schema::Result::Software::Building -> software_buildings
     MyApp::Schema::Result::LonelyPerson -> lonely_people
     MyApp::DB::Result::FriendlyPerson -> friendly_people
     MyApp::DB::Result::Dog -> dogs

    Also, if you just want to be different, you can easily set up your own
    naming scheme. Just add a "gen_table" method to your candy subclass. The
    method gets passed the class name and the autotable version, which of
    course you may ignore. For example, one might just do the following:

     sub gen_table {
       my ($self, $class) = @_;

       $class =~ s/::/_/g;
       lc $class;
     }

    Which would tranform "MyApp::Schema::Result::Foo" into
    "myapp_schema_result_foo".

    Or maybe instead of using the standard "MyApp::Schema::Result" namespace
    you decided to be different and do "MyApp::DB::Table" or something silly
    like that. You could pre-process your class name so that the default
    "gen_table" will still work:

     sub gen_table {
       my $self = shift;
       my $class = $_[0];

       $class =~ s/::DB::Table::/::Schema::Result::/;
       return $self->next::method(@_);
     }

AUTHOR
    Arthur Axel "fREW" Schmidt <frioux+cpan@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Arthur Axel "fREW" Schmidt.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for DBIx-Class-Candy

0.002107  2014-10-30 19:53:58-05:00 America/Chicago
  - Fix `export_method_aliases` (Thanks Denis Ibaev! See GH#7)

0.002106  2014-10-17 15:26:43-05:00 America/Chicago
  - Fix bug causing infinite recursion when you redefine a package (Thanks
    Graham Knop for fix, and Wes Malone for bug report!)
  - Simplify sub installer by using Sub::Exporter more correctly (Thanks RJBS!)
  - Stop calling ->table many times per ResultClass (Good catch RJBS!)

0.002105  2014-04-12 08:51:04-05:00 America/Chicago
  - Fix warnings in test (thanks for the report Zefram!) (Fixes RT#91976)

0.002104  2013-05-13 09:29:37 CST6CDT
  - Fix stupid doc examples

0.002103  2012-12-23 17:42:36 CST6CDT
  - Lazily load deps for autotable
  - Remove HERE BE DRAGONS warning, this module is totally stable \o/

0.002102  2012-11-17 15:43:18 CST6CDT
  - Put MetaYAML back in dist

0.002101  2012-08-21 17:47:50 America/Chicago
  - remove VERSION from pod because yuck
  - fix repo url

0.002100  2012-03-13 20:31:45 America/Chicago
  - primary_column may be called multiple times to define a multicolumn pk

0.002001  2011-08-10 20:03:18 CST6CDT
  - autotable v1 now dies on unrecognized naming scheme
  - autotable v1 works with simply ::Result:: and not just ::Schema::Result::

0.002000  2011-03-09 12:03:50 CST6CDT
  - Add unique_column sugar
  - Allow Candy to automatically set table name
  - Allow Candy subclass to define default base and default perl version
  - Add missing docs for primary_column sugar

0.001006  2011-03-01 22:44:45 CST6CDT
  - Add primary_column sugar
  - Fix incorrectly named export (inflate_colum .= n)

0.001005  2010-12-25 10:06:15 CST6CDT
  - Initial sketches of "Moosey" API
  - Merry Christmas!!!

0.001004  2010-07-31 01:46:04 CST6CDT
  - Fix a bug that only appears in perl 5.8

0.001003  2010-07-29 20:20:14 CST6CDT
  - Fix more bugs, this time the are from inheriting from other
    other results
  - Make ::Exports actually work

0.001002  2010-07-24 01:21:49 CST6CDT
  - Fix bug uncovered by doing belongs_to twice in a class

0.001001  2010-07-21 17:48:14 CST6CDT
  - Don't require parent for tests

0.001000  2010-07-21 00:18:10 CST6CDT
  - Initial Release



Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.020.
use strict;
use warnings;



use ExtUtils::MakeMaker ;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Sugar for your favorite ORM, DBIx::Class",
  "AUTHOR" => "Arthur Axel \"fREW\" Schmidt <frioux+cpan\@gmail.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "DBIx-Class-Candy",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "DBIx::Class::Candy",
  "PREREQ_PM" => {
    "DBIx::Class" => "0.08123",
    "Lingua::EN::Inflect" => 0,
    "MRO::Compat" => "0.11",
    "String::CamelCase" => 0,
    "Sub::Exporter" => "0.982",
    "namespace::clean" => "0.18"
  },
  "TEST_REQUIRES" => {
    "Test::Deep" => 0,
    "Test::Fatal" => 0,
    "Test::More" => "0.94"
  },
  "VERSION" => "0.002107",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "DBIx::Class" => "0.08123",
  "Lingua::EN::Inflect" => 0,
  "MRO::Compat" => "0.11",
  "String::CamelCase" => 0,
  "Sub::Exporter" => "0.982",
  "Test::Deep" => 0,
  "Test::Fatal" => 0,
  "Test::More" => "0.94",
  "namespace::clean" => "0.18"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/DBIx/Class/Candy/Exports.pm
package DBIx::Class::Candy::Exports;

# ABSTRACT: Create sugar for your favorite ORM, DBIx::Class

use strict;
use warnings;

our $VERSION = '0.002107'; # VERSION

our %methods;
our %aliases;

sub export_methods        { $methods{scalar caller(0)} = $_[0] }
sub export_method_aliases { $aliases{scalar caller(0)} = $_[0] }

use Sub::Exporter -setup => {
   exports => [ qw(export_methods export_method_aliases) ],
   groups  => { default => [ qw(export_methods export_method_aliases) ] },
};

1;

__END__

=pod

=head1 NAME

DBIx::Class::Candy::Exports - Create sugar for your favorite ORM, DBIx::Class

=head1 SYNOPSIS

 package DBIx::Class::Widget;

 sub create_a_widget { ... }

 # so you don't depend on ::Candy
 eval {
   require DBIx::Class::Candy::Exports;
   DBIx::Class::Candy::Exports->import;
   export_methods ['create_a_widget'];
   export_method_aliases {
     widget => 'create_a_widget'
   };
 }

 1;

The above will make it such that users of your component who use it with
L<DBIx::Class::Candy> will have the methods you designate exported into
their namespace.

=head1 DESCRIPTION

The whole point of this module is to make sugar a first class citizen in
the component world that dominates L<DBIx::Class>.  I make enough components
and like this sugar idea enough that I want to be able to have both at the
same time.

=head1 IMPORTED SUBROUTINES

=head2 export_methods

 export_methods [qw( foo bar baz )];

Use this subroutine to define methods that get exported as subroutines of the
same name.

=head2 export_method_aliases

 export_method_aliases {
   old_method_name => 'new_sub_name',
 };

Use this subroutine to define methods that get exported as subroutines of a
different name.

=head1 AUTHOR

Arthur Axel "fREW" Schmidt <frioux+cpan@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Arthur Axel "fREW" Schmidt.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/DBIx/Class/Candy.pm
package DBIx::Class::Candy;

use strict;
use warnings;

our $VERSION = '0.002107'; # VERSION

use namespace::clean;
require DBIx::Class::Candy::Exports;
use MRO::Compat;
use Sub::Exporter 'build_exporter';
use Carp 'croak';

# ABSTRACT: Sugar for your favorite ORM, DBIx::Class

my %aliases = (
   column            => 'add_columns',
   primary_key       => 'set_primary_key',
   unique_constraint => 'add_unique_constraint',
   relationship      => 'add_relationship',
);

my @methods = qw(
   resultset_class
   resultset_attributes
   remove_columns
   remove_column
   table
   source_name

   inflate_column

   belongs_to
   has_many
   might_have
   has_one
   many_to_many

   sequence
);

sub base { return $_[1] || 'DBIx::Class::Core' }

sub perl_version { return $_[1] }

sub autotable { $_[1] }

sub gen_table {
   my ( $self, $class, $version ) = @_;
   if ($version == 1) {
      if (my ( $part ) = $class =~ /(?:::Schema)?::Result::(.+)$/) {
         require Lingua::EN::Inflect;
         require String::CamelCase;
         $part =~ s/:://g;
         $part = String::CamelCase::decamelize($part);
         return join q{_}, split /\s+/,
            Lingua::EN::Inflect::PL(join q{ }, split /_/, $part);
      } else {
         croak 'unrecognized naming scheme!'
      }
   }
}

sub import {
   my $self = shift;

   my $inheritor = caller(0);
   my $args         = $self->parse_arguments(\@_);
   my $perl_version = $self->perl_version($args->{perl_version});
   my @rest         = @{$args->{rest}};

   $self->set_base($inheritor, $args->{base});
   $inheritor->load_components(@{$args->{components}});
   my @custom_methods;
   my %custom_aliases;
   {
      my @custom = $self->gen_custom_imports($inheritor);
      @custom_methods = @{$custom[0]};
      %custom_aliases = %{$custom[1]};
   }

   my $set_table = sub {};
   if (my $v = $self->autotable($args->{autotable})) {
     my $table_name = $self->gen_table($inheritor, $v);
     my $ran = 0;
     $set_table = sub { $inheritor->table($table_name) unless $ran++ }
   }
   @_ = ($self, @rest);
   my $import = build_exporter({
      exports => [
         has_column => $self->gen_has_column($inheritor, $set_table),
         primary_column => $self->gen_primary_column($inheritor, $set_table),
         unique_column => $self->gen_unique_column($inheritor, $set_table),
         (map { $_ => $self->gen_proxy($inheritor, $set_table) } @methods, @custom_methods),
         (map { $_ => $self->gen_rename_proxy($inheritor, $set_table, %aliases, %custom_aliases) }
            keys %aliases, keys %custom_aliases),
      ],
      groups  => {
         default => [
            qw(has_column primary_column unique_column), @methods, @custom_methods, keys %aliases, keys %custom_aliases
         ],
      },
      installer  => $self->installer,
      collectors => [
         INIT => $self->gen_INIT($perl_version, \%custom_aliases, \@custom_methods, $inheritor),
      ],
   });

   goto $import
}

sub gen_custom_imports {
  my ($self, $inheritor) = @_;
  my @methods;
  my %aliases;
  for (@{mro::get_linear_isa($inheritor)}) {
    if (my $a = $DBIx::Class::Candy::Exports::aliases{$_}) {
      %aliases = (%aliases, %$a)
    }
    if (my $m = $DBIx::Class::Candy::Exports::methods{$_}) {
      @methods = (@methods, @$m)
    }
  }
  return(\@methods, \%aliases)
}

sub parse_arguments {
  my $self = shift;
  my @args = @{shift @_};

  my $skipnext;
  my $base;
  my @rest;
  my $perl_version = undef;
  my $components   = [];
  my $autotable = 0;
  for my $idx ( 0 .. $#args ) {
    my $val = $args[$idx];

    next unless defined $val;
    if ($skipnext) {
      $skipnext--;
      next;
    }

    if ( $val eq '-base' ) {
      $base = $args[$idx + 1];
      $skipnext = 1;
    } elsif ( $val eq '-autotable' ) {
      $autotable = ord $args[$idx + 1];
      $skipnext = 1;
    } elsif ( $val eq '-perl5' ) {
      $perl_version = ord $args[$idx + 1];
      $skipnext = 1;
    } elsif ( $val eq '-components' ) {
      $components = $args[$idx + 1];
      $skipnext = 1;
    } else {
      push @rest, $val;
    }
  }

  return {
    autotable    => $autotable,
    base         => $base,
    perl_version => $perl_version,
    components   => $components,
    rest         => \@rest,
  };
}

sub gen_primary_column {
  my ($self, $inheritor, $set_table) = @_;
  sub {
    my $i = $inheritor;
    sub {
      my $column = shift;
      my $info   = shift;
      $set_table->();
      $i->add_columns($column => $info);
      $i->set_primary_key($i->primary_columns, $column);
    }
  }
}

sub gen_unique_column {
  my ($self, $inheritor, $set_table) = @_;
  sub {
    my $i = $inheritor;
    sub {
      my $column = shift;
      my $info   = shift;
      $set_table->();
      $i->add_columns($column => $info);
      $i->add_unique_constraint([ $column ]);
    }
  }
}

sub gen_has_column {
  my ($self, $inheritor, $set_table) = @_;
  sub {
    my $i = $inheritor;
    sub {
      my $column = shift;
      $set_table->();
      $i->add_columns($column => { @_ })
    }
  }
}

sub gen_rename_proxy {
  my ($self, $inheritor, $set_table, %aliases) = @_;
  sub {
    my ($class, $name) = @_;
    my $meth = $aliases{$name};
    my $i = $inheritor;
    sub { $set_table->(); $i->$meth(@_) }
  }
}

sub gen_proxy {
  my ($self, $inheritor, $set_table) = @_;
  sub {
    my ($class, $name) = @_;
    my $i = $inheritor;
    sub { $set_table->(); $i->$name(@_) }
  }
}

sub installer {
  my ($self) = @_;
  sub {
    Sub::Exporter::default_installer @_;
    my %subs = @{ $_[1] };
    namespace::clean->import( -cleanee => $_[0]{into}, keys %subs )
  }
}

sub set_base {
   my ($self, $inheritor, $base) = @_;

   # inlined from parent.pm
   for ( my @useless = $self->base($base) ) {
      s{::|'}{/}g;
      require "$_.pm"; # dies if the file is not found
   }

   {
      no strict 'refs';
      # This is more efficient than push for the new MRO
      # at least until the new MRO is fixed
      @{"$inheritor\::ISA"} = (@{"$inheritor\::ISA"} , $self->base($base));
   }
}

sub gen_INIT {
  my ($self, $perl_version, $custom_aliases, $custom_methods, $inheritor) = @_;
  sub {
    my $orig = $_[1]->{import_args};
    $_[1]->{import_args} = [];
    %$custom_aliases = ();
    @$custom_methods = ();

    if ($perl_version) {
       require feature;
       feature->import(":5.$perl_version")
    }

    strict->import;
    warnings->import;

    1;
  }
}

1;

__END__

=pod

=head1 NAME

DBIx::Class::Candy - Sugar for your favorite ORM, DBIx::Class

=head1 SYNOPSIS

 package MyApp::Schema::Result::Artist;

 use DBIx::Class::Candy -autotable => v1;

 primary_column id => {
   data_type => 'int',
   is_auto_increment => 1,
 };

 column name => {
   data_type => 'varchar',
   size => 25,
   is_nullable => 1,
 };

 has_many albums => 'A::Schema::Result::Album', 'artist_id';

 1;

=head1 DESCRIPTION

C<DBIx::Class::Candy> is a simple sugar layer for definition of
L<DBIx::Class> results.  Note that it may later be expanded to add sugar
for more C<DBIx::Class> related things.  By default C<DBIx::Class::Candy>:

=over

=item *

turns on strict and warnings

=item *

sets your parent class

=item *

exports a bunch of the package methods that you normally use to define your
L<DBIx::Class> results

=item *

makes a few aliases to make some of the original method names a shorter or
more clear

=item *

defines very few new subroutines that transform the arguments passed to them

=back

It assumes a L<DBIx::Class::Core>-like API, but you can tailor it to suit
your needs.

=head1 IMPORT OPTIONS

See L</SETTING DEFAULT IMPORT OPTIONS> for information on setting these schema wide.

=head2 -base

 use DBIx::Class::Candy -base => 'MyApp::Schema::Result';

The first thing you can do to customize your usage of C<DBIx::Class::Candy>
is change the parent class.  Do that by using the C<-base> import option.

=head2 -autotable

 use DBIx::Class::Candy -autotable => v1;

Don't waste your precious keystrokes typing C<< table 'buildings' >>, let
C<DBIx::Class::Candy> do that for you!  See L<AUTOTABLE VERSIONS> for what the
existing versions will generate for you.

=head2 -components

 use DBIx::Class::Candy -components => ['FilterColumn'];

C<DBIx::Class::Candy> allows you to set which components you are using at
import time so that the components can define their own sugar to export as
well.  See L<DBIx::Class::Candy::Exports> for details on how that works.

=head2 -perl5

 use DBIx::Class::Candy -perl5 => v10;

I love the new features in Perl 5.10 and 5.12, so I felt that it would be
nice to remove the boiler plate of doing C<< use feature ':5.10' >> and
add it to my sugar importer.  Feel free not to use this.

=head1 IMPORTED SUBROUTINES

Most of the imported subroutines are the same as what you get when you use
the normal interface for result definition: they have the same names and take
the same arguments.  In general write the code the way you normally would,
leaving out the C<< __PACKAGE__-> >> part.  The following are methods that
are exported with the same name and arguments:

 belongs_to
 has_many
 has_one
 inflate_column
 many_to_many
 might_have
 remove_column
 remove_columns
 resultset_attributes
 resultset_class
 sequence
 source_name
 table

There are some exceptions though, which brings us to:

=head1 IMPORTED ALIASES

These are merely renamed versions of the functions you know and love.  The idea is
to make your result classes a tiny bit prettier by aliasing some methods.
If you know your C<DBIx::Class> API you noticed that in the L</SYNOPSIS> I used C<column>
instead of C<add_columns> and C<primary_key> instead of C<set_primary_key>.  The old
versions work, this is just nicer.  A list of aliases are as follows:

 column            => 'add_columns',
 primary_key       => 'set_primary_key',
 unique_constraint => 'add_unique_constraint',
 relationship      => 'add_relationship',

=head1 SETTING DEFAULT IMPORT OPTIONS

Eventually you will get tired of writing the following in every single one of
your results:

 use DBIx::Class::Candy
   -base      => 'MyApp::Schema::Result',
   -perl5     => v12,
   -autotable => v1;

You can set all of these for your whole schema if you define your own C<Candy>
subclass as follows:

 package MyApp::Schema::Candy;

 use base 'DBIx::Class::Candy';

 sub base { $_[1] || 'MyApp::Schema::Result' }
 sub perl_version { 12 }
 sub autotable { 1 }

Note the C<< $_[1] || >> in C<base>.  All of these methods are passed the
values passed in from the arguments to the subclass, so you can either throw
them away, honor them, die on usage, or whatever.  To be clear, if you define
your subclass, and someone uses it as follows:

 use MyApp::Schema::Candy -base => 'MyApp::Schema::Result', -perl5 => v18, -autotable => v1;

Your C<base> method will get C<MyApp::Schema::Result>, your
C<perl_version> will get C<18>, and your C<autotable> will get
C<1>.

=head1 SECONDARY API

=head2 has_column

There is currently a single "transformer" for C<add_columns>, so that
people used to the L<Moose> api will feel more at home.  Note that this B<may>
go into a "Candy Component" at some point.

Example usage:

 has_column foo => (
   data_type => 'varchar',
   size => 25,
   is_nullable => 1,
 );

=head2 primary_column

Another handy little feature that allows you to define a column and set it as
the primary key in a single call:

 primary_column id => {
   data_type => 'int',
   is_auto_increment => 1,
 };

If your table has multiple columns in its primary key, merely call this method
for each column:

 primary_column person_id => { data_type => 'int' };
 primary_column friend_id => { data_type => 'int' };

=head2 unique_column

This allows you to define a column and set it as unique in a single call:

 unique_column name => {
   data_type => 'varchar',
   size => 30,
 };

=head1 AUTOTABLE VERSIONS

Currently there is a single version, C<v1>, which looks at your class name,
grabs everything after C<::Schema::Result::> (or C<::Result::>), removes the
C<::>'s, converts it to underscores instead of camel-case, and pluralizes it.
Here are some examples if that's not clear:

 MyApp::Schema::Result::Cat -> cats
 MyApp::Schema::Result::Software::Building -> software_buildings
 MyApp::Schema::Result::LonelyPerson -> lonely_people
 MyApp::DB::Result::FriendlyPerson -> friendly_people
 MyApp::DB::Result::Dog -> dogs

Also, if you just want to be different, you can easily set up your own naming
scheme.  Just add a C<gen_table> method to your candy subclass.  The method
gets passed the class name and the autotable version, which of course you may
ignore.  For example, one might just do the following:

 sub gen_table {
   my ($self, $class) = @_;

   $class =~ s/::/_/g;
   lc $class;
 }

Which would tranform C<MyApp::Schema::Result::Foo> into
C<myapp_schema_result_foo>.

Or maybe instead of using the standard C<MyApp::Schema::Result> namespace you
decided to be different and do C<MyApp::DB::Table> or something silly like that.
You could pre-process your class name so that the default C<gen_table> will
still work:

 sub gen_table {
   my $self = shift;
   my $class = $_[0];

   $class =~ s/::DB::Table::/::Schema::Result::/;
   return $self->next::method(@_);
 }

=head1 AUTHOR

Arthur Axel "fREW" Schmidt <frioux+cpan@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Arthur Axel "fREW" Schmidt.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
